pkg/virtctl/imageupload/imageupload.go

```diff
@@ -0,0 +1,355 @@
+
+	req.Header.Add("Authorization", "Bearer "+token)
+	req.Header.Add("Content-Type", "application/octet-stream")
+
+	fmt.Println()
+	bar.Start()
+
+	resp, err := client.Do(req)
+
+	bar.Finish()
+	fmt.Println()
+
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Unexpected return value %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func getUploadToken(client cdiClientset.Interface, namespace, name string) (string, error) {
+	request := &uploadcdiv1.UploadTokenRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "token-for-virtctl",
+		},
+		Spec: uploadcdiv1.UploadTokenRequestSpec{
+			PvcName: name,
+		},
+	}
+
+	response, err := client.UploadV1alpha1().UploadTokenRequests(namespace).Create(request)
+	if err != nil {
+		return "", err
+	}
+
+	return response.Status.Token, nil
+}
+
+func waitUploadPodRunning(client kubernetes.Interface, namespace, name string, interval, timeout time.Duration) error {
+	serviceName := "cdi-upload-" + name
+	loggedStatus := false
+
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		pvc, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		endpoints, err := client.CoreV1().Endpoints(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		podPhase, _ := pvc.Annotations[PodPhaseAnnotation]
+
+		done := (podPhase == string(v1.PodRunning)) && (len(endpoints.Subsets) > 0)
+
+		if !done && !loggedStatus {
+			fmt.Printf("Waiting for PVC %s upload pod to be running...\n", name)
+			loggedStatus = true
+		}
+
+		if done && loggedStatus {
+			// be really sure
+			time.Sleep(interval)
+			fmt.Printf("Pod now running\n")
+		}
+
+		return done, nil
+	})
+
+	return err
+}
+
+func createUploadPVC(client kubernetes.Interface, namespace, name, size, storageClass, accessMode string) (*v1.PersistentVolumeClaim, error) {
+	quantity, err := resource.ParseQuantity(size)
+	if err != nil {
+		return nil, err
+	}
+
+	pvc := &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Annotations: map[string]string{
+				uploadRequestAnnotation: "",
+			},
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: quantity,
+				},
+			},
+		},
+	}
+
+	if storageClass != "" {
+		pvc.Spec.StorageClassName = &storageClass
+	}
+
+	if accessMode != "" {
+		pvc.Spec.AccessModes = []v1.PersistentVolumeAccessMode{v1.PersistentVolumeAccessMode(accessMode)}
+	}
+
+	pvc, err = client.CoreV1().PersistentVolumeClaims(namespace).Create(pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	return pvc, nil
+}
+
+func getUploadPVC(client kubernetes.Interface, namespace, name string, shouldExist bool) (*v1.PersistentVolumeClaim, error) {
```

Comment:
Can you remove the `shouldExist` parameter? Let's not manipulate a PVC in a function starting with `get`.

Reply:
changed function to be getAndValidate().  Doesn't modify PVC.  But needs should exist flag for validation.